### ISSUE TITLE
Kernel/Interrupt Kernel does not terminate stuck subprocesses in the notebook

### ISSUE THREAD

**Issue by pfmoore** [ipython/ipython#3400]
When a subprocess is run from the notebook, if it gets stuck the kernel will get locked waiting for it. Selecting Kernel/Interrupt from the menu does not terminate the subprocess, but rather leaves the kernel in an unstable, "partially locked" state, where other cells do not execute. The only resolution is to restart the kernel.

This occurred for me on Windows - I do not know if it also happens on Unix.

To demonstrate, start a notebook and enter !python in a cell. The process will lock as it is waiting for interactive input. As there is no way to provide that input, the kernel must be restarted to continue.

**Comment by wmayner** [comment]
I think I just got bitten by this and I'll need to restart the kernel, meaning I've just lost a lot of data…

I was using `pdb` to debug a function. I re-ran the cell without first quitting `pdb`, and now I can't interrupt anything.

Here's a minimal example that reproduces this:
```python
def test():
    import pdb; pdb.set_trace()  # XXX BREAKPOINT
    return 0

test()
```
Run this cell twice in a row.

**Comment by lancekrogers** [comment]
This same issue happens for me in Unix as well word for word.  

"When a subprocess is run from the notebook, if it gets stuck the kernel will get locked waiting for it. Selecting Kernel/Interrupt from the menu does not terminate the subprocess, but rather leaves the kernel in an unstable, "partially locked" state, where other cells do not execute. The only resolution is to restart the kernel."

**Comment by nealmcb** [comment]
Thanks for the nice example of a pdb hang, wmayner. But sInce pdb doesn't run in a subprocess, I opened a separate issue for pdb: #10516

**Comment by JulesGM** [comment]
Printing too much data, let's say a gigantic numpy array, can make the kernel completely unresponsive and impossible to to terminate

**Comment by rajulah** [comment]
Has a solution been found for this issue yet? i just ran a machine learning model that took 14hr to complete and now my kernel is stuck and doesnt execute cells. if i restart, i have to run the model again for 14hrs. So is there any solution?

**Comment by JulesGM** [comment]
haven't tried it, but this seems like it could help: http://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/limit_output/readme.html

**Comment by takluyver** [comment]
If a specific subprocess has got stuck, you can probably find it in the task manager and forcibly kill it that way. Hopefully that lets the kernel continue.

**Comment by patricktokeeffe** [comment]
I'm dealing with a stuck notebook too: interrupt, restart, reconnect - none of them do anything. The `[*]` indicators remain next to cells as if they are queued to run but no cells get executed. 

The behavior began after running a cell containing:
```
filedir = "20161214_rooftest"

!ls -RC $filedir
```

Which is strange because I have analogous cells elsewhere that run successfully. I'm not sure how/if `ls` could get stuck but otherwise my situation seems to match this issue.

**Comment by ashishanand7** [comment]
Is there any solution to this . Kernal cannot be interrupted .
For me it's happening with GridSearchCV in sklearn .

**Comment by ahmedrao** [comment]
There was a process named conda.exe in Task manager. I killed that process and I was successfully able to interrupt the kernel

**Comment by IMBurbank** [comment]
Interrupt is still broken. I have to restart and reload my imports every time.

**Comment by metya** [comment]
same problem in jupyter lab on python 3.7 kernel

**Comment by CathyQian** [comment]
same problem in Jupyter Notebook and I can't find the process named conda.exe in Task manager. Any updates on the solution yet?

**Comment by esha-sg** [comment]
**Not a solution**
Sometimes trying to reconnect to the kernel helps in this case

**Comment by ambareeshsrja16** [comment]
Observing the same, in Windows 10

**Comment by arianccbasile** [comment]
Did anyone succeed on that? I am getting crazy

**Comment by kak-to-tak** [comment]
> There was a process named conda.exe in Task manager. I killed that process and I was successfully able to interrupt the kernel

How????

**Comment by rudaoshi** [comment]
This problem has existed for six years and still no solution.

**Comment by HamdiTarek** [comment]
> This problem has existed for six years and still no solution.

six years without any solution, just restart the kernel

**Comment by Stuj79** [comment]
Having the same problem increasingly frequently, almost to the point where the notebooks are becoming unusable which is a real shame. On Anaconda 3.7 and the cells just hang with the asterisk, and I am unable to interrupt the kernel.

**Comment by vinklibrary** [comment]
Mark Same Issue

**Comment by ChrisPalmerNZ** [comment]
Have always had this problem especially with dbg and input. 
Windows 10; Notebook server 5.7.8; Python 3.6.6.; Conda 4.7.5
Have learned that I basically cannot reliably debug Notebooks :(

**Comment by SangamSwadiK** [comment]
yep, the problem still exists. Is there any way to over come this ?? I dont want to run my notebook all over again , because it takes too long to get to where I'm !!

**Comment by louismartin** [comment]
Up!
This problem has been a pain for me for years now every time I use pdb and forget to quit before I re-run the cell.

**Comment by louismartin** [comment]
I created a bounty on BountySource. Maybe this will finally be fixed if we can gather enough money.
https://www.bountysource.com/issues/44958889-hang-after-running-pdb-in-a-cell-kernel-interrupt-doesn-t-help

**Comment by itamarst** [comment]
For the process issue specifically, on Windows specifically, here's a theory (still untested):

1. Process is run via `IPython.utils._process_win32.system`, which calls `_system_body`, which calls `p.wait()` on the `subprocess.Popen` object.
2. Windows `subprocess.Popen.wait()` has a known issue where it is not interruptible: https://bugs.python.org/issue28168

If that's the cause, switching to busy looping every 100ms or so would probably make it interruptible, or if not then taking the approach in the patch.

**Comment by nealmcb** [comment]
Thank you @Carreau!

**Comment by ChrisPalmerNZ** [comment]
Thanks @Carreau! When will this find its way into a general release, and does it mean that we will then be able to use the Interrupt Kernel button sucessfully?

**Comment by Carreau** [comment]
I'll likely do a 7.13 tomorrow. It might fix the interrupt button.

**Comment by Arpit-Gole** [comment]
Hey @Carreau 
I am facing this issue when I am trying to interrupt an ongoing cell execution, interrupt goes on forever and at last I have to restart.

So in order to demonstrate, as @wmayner suggested a way to replicate the issue. I have attached a few screenshots for the same.
![pyt1](https://user-images.githubusercontent.com/37917285/80300937-4dfa3280-87be-11ea-85cc-96fb1e09a84e.png)

**Comment by itamarst** [comment]
@Arpit-Gole pdb is its own specific issue; I'm hoping to get that fixed soon too: https://github.com/ipython/ipython/issues/10516

**Comment by Arpit-Gole** [comment]
@itamarst I am training a model as follows :

`forest_clf = RandomForestClassifier()
cross_val_score(forest_clf, X_train, y_train, cv=3, scoring='accuracy', verbose=10, n_jobs=-1)`

Now I know it is bound to take time-based on my dataset. But say for whatever reason I choose to stop the processing in half-way by pressing **Kernel>Interrupt Kernel**.
Ideally, it should interrupt but it takes forever to stop.
Now I don't want to restart because all my progress will be gone.

Please Help!

**Comment by Carreau** [comment]
If what you are trying to interrupt is implemented in C then there is nothing to do. It's up to the library you use to handle sigint.

**Comment by julianportelli** [comment]
> 
> 
> @itamarst I am training a model as follows :
> 
> `forest_clf = RandomForestClassifier() cross_val_score(forest_clf, X_train, y_train, cv=3, scoring='accuracy', verbose=10, n_jobs=-1)`
> 
> Now I know it is bound to take time-based on my dataset. But say for whatever reason I choose to stop the processing in half-way by pressing **Kernel>Interrupt Kernel**.
> Ideally, it should interrupt but it takes forever to stop.
> Now I don't want to restart because all my progress will be gone.
> 
> Please Help!

Same for me. Interrupting the Kernel does nothing. I am using GridSearchCV on sklearn.SVR  and I can't stop it to change the parameters. Any decent solution or is this not fixable?

**Comment by jvschoen** [comment]
Just wanted to add a reproducible example running in Jupyterlab:

I run into this sometimes too... Here is a reproduceable example from jupyer lab:
## LOAD DATA
```
import requests
import pandas as pd

url='https://raw.githubusercontent.com/numenta/NAB/master/data/realKnownCause/nyc_taxi.csv'
r = requests.get(url, allow_redirects=True)
        with open('data/nyc_taxi.csv', 'wb') as f:
            f.write(r.content)
df_taxi = (
        pd.read_csv('data/nyc_taxi.csv')
        .assign(timestamp=lambda x: pd.to_datetime(x.timestamp))
)

df_train = df_taxi.iloc[:5000]
temp_train = df_train.set_index('timestamp')
```
## Run Grid Search: THIS CANNOT BE INTERRUPTED
```
import itertools
#set parameter range
p = range(0,3)
q = range(1,3)
d = range(1,2)
s = [24,48]

# list of all parameter combos
pdq = list(itertools.product(p, d, q))
seasonal_pdq = list(itertools.product(p, d, q, s))
# SARIMA model pipeline
for param in pdq:
    for param_seasonal in seasonal_pdq:
        try:
            mod = sm.tsa.statespace.SARIMAX(temp_train[:240],
                                            order=param,
                                            seasonal_order=param_seasonal)
            
            results = mod.fit(max_iter = 50, method = 'powell')
            
            print('SARIMA{},{} - AIC:{}'.format(param, param_seasonal, results.aic))
        except as e:
            print(e)
            continue
```

Is there any advice?

**Comment by dawnset** [comment]
run into this problem three times this afternoon, reminds me of the good old days when i was still using urllib.
thought its on urllib, cause there is no response to my request.
I was working but coding, I have to find a solution but a answer. So I store every variable to local file.
really don't want to see that happen again and again.

**Comment by Crispy13** [comment]
I am facing the same issue when using tensorflow and gpu for training deep learning model.

**Comment by matija2209** [comment]
Run into this with time.sleep and requests

**Comment by mbrad092** [comment]
Also having this issue with time.sleep requests on Windows, but runs fine on Mac OS X

**Comment by not-Ian** [comment]
Having this issue with ThreadPoolExecutor... Something like this:

```
numberOfImageGatherers = 2

with concurrent.futures.ThreadPoolExecutor(max_workers=numberOfImageGatherers + 1) as executor:
        futures = []

        for imageGatherer in range(numberOfImageGatherers):
            imageDataGatherer = ImageDataGatherer(batch_size)
            futures.append(executor.submit(imageDataGatherer.gatherImageData, pipeline))

        modelTrainingConsumer = ModelTrainingConsumer(vae, plot_losses)    

        futures.append(executor.submit(modelTrainingConsumer.trainModel, pipeline))

        concurrent.futures.wait(futures)
```

Only way to interrupt is to restart kernel... very frustrating

**Comment by TV4Fun** [comment]
This is still happening. I would suggest re-opening this issue. Seeing it in a NumPy-heavy tight neural network training loop on Windows 10.
```
jupyter core     : 4.7.1
jupyter-notebook : 6.2.0
qtconsole        : 4.7.7
ipython          : 7.20.0
ipykernel        : 5.3.4
jupyter client   : 6.1.7
jupyter lab      : 2.2.6
nbconvert        : 6.0.7
ipywidgets       : 7.6.3
nbformat         : 5.1.2
traitlets        : 5.0.5
```
Is there anything I need to upgrade?

**Comment by Esesna** [comment]
Возникает такая же проблема, и самое интересное в случайных участках кода при использование B0 remote api для coppelia sim. При это если я использую Publisher и Subscriber

**Comment by jasperSha** [comment]
So I implemented this as a fix, this was done on Linux. I think it's just a bandaid fix but it manages to keep the current variables et al. It's essentially taking over and manually sending the SIGINT call to interrupt the script when telling jupyter to interrupt. I usually use this for scripts that cannot be interrupted for ctrl+c, figured it might work here, and seems to have worked thus far, let me know if it doesn't!

```python
import signal, sys
terminate = False

def signal_handling(signum, frame):
    global terminate
    terminate = True
signal.signal(signal.SIGINT, signal_handling)

some_data = []
for doing_work in however_many_epochs():
    if terminate:
        break
    some_data.append(do_work)
```
after this, if you hit ii or ctrl+c or click the interrupt kernel button on jupyter it will stop the process, and some_data will still have whatever data it had stored up to that point.

**Comment by hamitaksln** [comment]
I found this solution to stop cell when I work with time sleep or requests:
![keyboard interrupt](https://i.gyazo.com/e75bb23c6c08b92b5fa28c02845c352b.gif)
```import time
for i in range(20):
    try:
        print(i)
        time.sleep(1)
    except KeyboardInterrupt:
        print("Stopping...")
        break
```

**Comment by thistlillo** [comment]
Are there any news on this issue? I am working remotely in a Kubernetes pod and, when this occurs, I am completely locked out of the machine. I cannot even open a shell.

**Comment by MrMino** [comment]
@thistlillo your issue is most likely unrelated to this one. In any case, we would need more information than what you've provided, like: version of IPython and Python you're using, whether you are using Jupyter Notebook or Jupyter Lab, versions of those, etc. And, most importantly, a minimalized set of steps required to reproduce the issue.

If you can provide this info, please create a new issue and include it.

**Comment by Ben-Epstein** [comment]
@MrMino I'm still experiencing this issue. I have a loop that simply makes a `request.get` and parses the response. I cannot interrupt the kernel
```
jupyter                  1.0.0
jupyter_client           8.2.0
jupyter-console          6.6.3
jupyter_core             5.3.0
jupyter-events           0.6.3
jupyter_server           2.5.0
jupyter_server_terminals 0.4.4
jupyterlab-pygments      0.2.2
jupyterlab-widgets       3.0.7
```

**Comment by UmarZein** [comment]
I ran into this issue today on my linux laptop. Although I have not found a way to interrupt the kernel, I was able to save my variables by injecting a `pickle.dumps` using `pyrasite`. Thanks to https://stackoverflow.com/a/59124617/18638036

**Comment by Shubhang9** [comment]
Hey guys! I also had the same issue, but it was a silly mistake on my end.

Avoid using try without adding specific exceptions in your except block you are expecting your code to throw. A generic try looks like this -

```
try:
    #code here
except:
    #exception handling
```

It seems like this code would also catches your keyboard Interrupts/ kernel interrupts which registers as KeyboardInterrupt exception.

 I thought it might help some unfortunate souls if I share it here.

**Comment by ericdrosas87** [comment]
Just tossing my 🎩 into the ring that I too am experiencing this issue

**Comment by rajhlinux** [comment]
Its 2025... getting same error, endless cycle of Interrupts kernel dialog while on the children Jupyter notebook to teach adult "programmers" how to program.

Edit:

So need to update VS code and then restart... Jupyter (stuff used to teach kids how to program for "software engineers" 🤣) then doesn't throw the endless "Interrupts kernel dialog". Programmers will get dumb when things are taught via Jupyter, it's so unfortunate.
